### PR TITLE
Release version 1.4.0

### PR DESCRIPTION
**Pull request by tiutiunnykleasoft**
* Implemented supporting `update` method for such collections :
  * Order Lines.
  * Abstract Collection.
  * Additional Addresses.
  * Transactions.
  * Phone Numbers.
* Implemented supporting `update` method for such entities : 
  * Line.
  * Address.
  * Customer.
  * Extra.
  * Payment Method Details.
  * Transaction.
* Implemented `update` method to `MultiFieldEntityInterface`.
* Implemented properties: 
  * `RawCost`.
  * `Percentage`.
  * `VatPercentage`
* Implemented tests: 
  * `RawCostTest`.
  * `AmountTest`.
  * `VatPercentageTest`.
  * `UpdateOrderTest`.
* Updated and Simplified:
  * `AdditionallAddressesTest`.
  * `CustomerTtest`.
* Updated  `get` calls for such entities:
  * Order.
  * Customer.
  * Transaction.
* Redesigned `Amount` property to expect two possible variants in constructor `RawCost` property or value in cents.
* Redesigned `fromArray` method to be static.
* Updated tests to use `fromArray` as a static method.
* Updated `fromArray` method.
* Eliminated `ValueIncentsInterface`.

### CODE REVIEW

File: src/Entities/Order.php [hunk 1]
```diff
@@ -56,59 +46,39 @@ public function __construct(
...
-    public function getMerchantOrderId(): string
-    {
-        return $this->merchantOrderId->get();
-    }
-
-    public function getExtra(): array
-    {
-        return $this->extra?->toArray();
-    }
-
```
Comment (BohdanLeasoft) [inline src/Entities/Order.php:108]: Are functions like getMerchantOrderId or getWebhookUrl moved somewhere? Or they were just deleted?

File: src/Tests/TransactionTest.php [hunk 2]
```diff
@@ -419,7 +410,7 @@ public function test_additional_property_reason()
 
     public function test_additional_property_is_capturable()
```
Comment (BohdanLeasoft) [inline src/Tests/TransactionTest.php:411]: Functions like  test_additional_property_is_capturable, test_method_is_capturable, test_additional_property_order_id
Looks similar. I offer the following option to reduce repetitions (it should be tested):

````
public function test_additional_property_is_capturable()
{
    $real = $this->getRealForTest(["is_capturable" => true]);
    $this->assertToArrayTest(
        $real,
        true,
        "is_capturable"
    );    
}

public function test_method_is_capturable()
{
    $real = $this->getRealForTest(["is_capturable" => true]);
    $this->assertIsCapturableTest(
        $real,
        true
    );
}

public function test_additional_property_order_id()
{    
    $real = $this->getRealForTest(["order_id" => "123"]);
    $this->assertToArrayTest(
        $real,
        "123",
        "order_id"
    );   
}

public function test_additional_property_channel_valid()
{
    $real = $this->getRealForTest(["channel" => "ecom"]);
    $this->assertToArrayTest(
        $real,
        "ecom",
        "channel"
    );    
}

public function test_additional_property_project_type()
{
    $real = $this->getRealForTest(["project_type" => "ginger"]);
    $this->assertToArrayTest(
        $real,
        "ginger",
        "project_type"
    );
}

private function getRealForTest($mockedData)
{
    $real = $this->client->fromArray(
        $real = Client::fromArray(
            Transaction::class,
            array_merge(
                self::MOCK_DATA_FOR_TRANSACTION,
                $mockedData
            )
        );
   return $real;
}

private function assertToArrayTest($real, $expected, $realId)
{    
    return self::assertSame(
        $real->toArray()[$realId],
        $expected
    );
}

private function assertIsCapturableTest($real, $expected)
{
    self::assertSame(
        $real->isCapturable(),
        $expected
    );
}
```
What do you thinks about such changes?

**Review by BohdanLeasoft — APPROVED** [review]
Good job :+1:

File: src/Entities/Order.php (hunk 1 above)
Comment (tiutiunnykleasoft) [reply src/Entities/Order.php:108]: Removed, because i don't see cases for usages. Do you see where it could be useful?

Comment (BohdanLeasoft) [reply src/Entities/Order.php:108]: MerchantOrderId could be used in some messages to customer. On success page or in emails.
Getting webhook url from order, probably, useless, I agree

Comment (tiutiunnykleasoft) [reply src/Entities/Order.php:108]: Ye, you might be right. I think I just missed up `merchant_order_id` property at all at this entity. Will implement in next commit.

File: src/Tests/TransactionTest.php (hunk 2 above)
Comment (tiutiunnykleasoft) [reply src/Tests/TransactionTest.php:411]: your suggestion is good but i think testing additional property is out of scope for now due to latest implemented method additional property filter

File: src/Tests/TransactionTest.php [hunk 3]
```diff
@@ -106,324 +90,6 @@ public function test_method_get_id()
         );
     }
 
-    public function test_additional_property_merchant_id()
```
Comment (BohdanLeasoft) [inline src/Tests/TransactionTest.php:118]: A lot of test deleted. Are they not needed at all? Will the percentage of code coverage by tests decrease?
It's nice that new additional property filter will check it, and completely replaces these tests

File: src/Client.php [hunk 4]
```diff
@@ -90,16 +89,16 @@ public function captureOrderTransaction(string $id): bool
     {
         $order = $this->getOrder(id: $id);
 
-        if ($order->getStatus() !== 'completed') {
+        if ($order->getStatus()->get() !== 'completed') {
             throw new InvalidOrderStatusException(
-                actual: $order->getStatus(), expected: 'completed'
+                actual: $order->getStatus()->get(), expected: 'completed'
```
Comment (Denis-Tsyganok-Leasoft) [inline src/Client.php:94]: I find `$order->getStatus()->get()` a bit alogical, I suggest to use:
1) `$order->status()->get()`
2) `$order->getStatus()`

File: src/Client.php [hunk 5]
```diff
@@ -90,16 +89,16 @@ public function captureOrderTransaction(string $id): bool
...
             );
         }
 
         try {
             $this->api_client->captureOrderTransaction(
                 orderId: $id,
-                transactionId: $order->getCurrentTransaction()->getId()
+                transactionId: $order->getCurrentTransaction()->getId()->get()
```
Comment (Denis-Tsyganok-Leasoft) [inline src/Client.php:101]: the same as above

**Review by Denis-Tsyganok-Leasoft — APPROVED** [review]
I've found only one moment, it's strange for me to see twice "get", but it's up to you 🙌

GJ

File: src/Client.php (hunk 4 above)
Comment (tiutiunnykleasoft) [reply src/Client.php:94]: > $order->status()->get()

I'm not sure that access to properties should be made public. All user needs are covered by one or another method of working with data.

Public access can harm the user himself.

**Pull request merged**